File: copilot-code/typescript/src/algorithms/boom-bust-scoring.ts
import { logger } from '../config';
import { BoomBustProfile } from '../types';

/**
 * Boom/bust scoring for player variance analysis
 * 
 * Tracks high-variance vs. consistent floor players
 * Recommends by draft stage (early = floor, late = boom)
 */

interface PlayerPerformance {
  playerId: string;
  weeklyScores: number[];
  gamesPlayed: number;
}

export function calculateBoomBustScore(
  performance: PlayerPerformance
): BoomBustProfile {
  try {
    const { playerId, weeklyScores } = performance;

    if (weeklyScores.length === 0) {
      logger.warn('No weekly scores for player', { playerId });
      return {
        player_id: playerId,
        variance_score: 0,
        floor_score: 0,
        ceiling_score: 0,
        consistency_rating: 'floor',
        recommended_draft_stage: 'mid',
      };
    }

    // Calculate statistical measures
    const mean = weeklyScores.reduce((sum, score) => sum + score, 0) / weeklyScores.length;
    const variance =
      weeklyScores.reduce((sum, score) => sum + Math.pow(score - mean, 2), 0) /
      weeklyScores.length;
    const stdDev = Math.sqrt(variance);

    // Normalize variance score (0-1 scale)
    const varianceScore = Math.min(stdDev / mean, 1.0);

    // Floor: 20th percentile of scores
    const sortedScores = [...weeklyScores].sort((a, b) => a - b);
    const floorIndex = Math.floor(sortedScores.length * 0.2);
    const floorScore = sortedScores[floorIndex];

    // Ceiling: 80th percentile of scores
    const ceilingIndex = Math.floor(sortedScores.length * 0.8);
    const ceilingScore = sortedScores[ceilingIndex];

    // Classify player
    let consistencyRating: 'boom' | 'bust' | 'floor' | 'ceiling';
    let recommendedStage: 'early' | 'mid' | 'late';

    if (varianceScore > 0.4) {
      // High variance
      if (mean > 15) {
        consistencyRating = 'boom'; // High-variance, high-scoring
        recommendedStage = 'mid';
      } else {
        consistencyRating = 'bust'; // High-variance, low-scoring
        recommendedStage = 'late';
      }
    } else {
      // Low variance (consistent)
      if (floorScore > 12) {
        consistencyRating = 'floor'; // High floor
        recommendedStage = 'early';
      } else {
        consistencyRating = 'ceiling'; // High ceiling but lower floor
        recommendedStage = 'mid';
      }
    }

    logger.debug('Boom/bust score calculated', {
      playerId,
      varianceScore,
      floorScore,
      ceilingScore,
      consistencyRating,
    });

    return {
      player_id: playerId,
      variance_score: varianceScore,
      floor_score: floorScore,
      ceiling_score: ceilingScore,
      consistency_rating: consistencyRating,
      recommended_draft_stage: recommendedStage,
    };
  } catch (error) {
    logger.error('Failed to calculate boom/bust score', {
      playerId: performance.playerId,
      error: error instanceof Error ? error.message : String(error),
    });
    throw error;
  }
}

/**
 * Get player recommendation based on draft stage
 * Early rounds: Prefer high floor
 * Late rounds: Prefer high ceiling/boom potential
 */
export function getRecommendationByStage(
  profiles: BoomBustProfile[],
  currentRound: number,
  totalRounds: number = 15
): BoomBustProfile[] {
  try {
    const stage =
      currentRound <= 5
        ? 'early'
        : currentRound >= 12
        ? 'late'
        : 'mid';

    logger.debug('Filtering by draft stage', { stage, currentRound });

    if (stage === 'early') {
      // Early rounds: Prefer consistent floor players
      return profiles
        .filter((p) => p.consistency_rating === 'floor')
        .sort((a, b) => b.floor_score - a.floor_score);
    } else if (stage === 'late') {
      // Late rounds: Prefer boom/upside players
      return profiles
        .filter(
          (p) =>
            p.consistency_rating === 'boom' ||
            p.consistency_rating === 'ceiling'
        )
        .sort((a, b) => b.ceiling_score - a.ceiling_score);
    } else {
      // Mid rounds: Balanced approach
      return profiles.sort(
        (a, b) =>
          b.floor_score +
          b.ceiling_score -
          (a.floor_score + a.ceiling_score)
      );
    }
  } catch (error) {
    logger.error('Failed to get recommendation by stage', {
      currentRound,
      error: error instanceof Error ? error.message : String(error),
    });
    throw error;
  }
}

/**
 * Calculate weekly performance from historical data
 * TODO: Query database for player's weekly scores
 */
export async function getPlayerPerformance(
  playerId: string,
  seasons: number[]
): Promise<PlayerPerformance> {
  try {
    // TODO: Implement database query
    // SELECT week, points FROM player_performances
    // WHERE player_id = $1 AND season IN ($2)
    // ORDER BY season, week

    logger.debug('Getting player performance', { playerId, seasons });

    // Placeholder
    return {
      playerId,
      weeklyScores: [],
      gamesPlayed: 0,
    };
  } catch (error) {
    logger.error('Failed to get player performance', {
      playerId,
      error: error instanceof Error ? error.message : String(error),
    });
    throw error;
  }
}

/**
 * Batch calculate boom/bust profiles for multiple players
 */
export async function calculateBatchBoomBust(
  playerIds: string[],
  seasons: number[]
): Promise<BoomBustProfile[]> {
  try {
    logger.info('Calculating boom/bust profiles', {
      playerCount: playerIds.length,
      seasons,
    });

    const profiles: BoomBustProfile[] = [];

    for (const playerId of playerIds) {
      const performance = await getPlayerPerformance(playerId, seasons);
      const profile = calculateBoomBustScore(performance);
      profiles.push(profile);
    }

    logger.info('Boom/bust profiles calculated', { count: profiles.length });
    return profiles;
  } catch (error) {
    logger.error('Failed to calculate batch boom/bust', {
      error: error instanceof Error ? error.message : String(error),
    });
    throw error;
  }
}

File: copilot-code/python/src/algorithms/boom_bust_scoring.py
"""Boom/bust scoring algorithm"""
import logging
from typing import List, Dict
from src.models.types import BoomBustProfile

logger = logging.getLogger(__name__)


def calculate_boom_bust_score(
    player_id: str,
    weekly_scores: List[float],
    boom_threshold: float = 20.0,
    bust_threshold: float = 5.0
) -> BoomBustProfile:
    """
    Calculate boom/bust profile based on weekly score variance
    
    Args:
        player_id: Player ID
        weekly_scores: List of weekly fantasy scores
        boom_threshold: Points threshold for boom week
        bust_threshold: Points threshold for bust week
    
    Returns:
        BoomBustProfile with variance metrics
    """
    try:
        if not weekly_scores:
            return BoomBustProfile(
                player_id=player_id,
                boom_rate=0.0,
                bust_rate=0.0,
                consistency_score=0.0,
                variance=0.0
            )
        
        # Calculate boom/bust rates
        boom_weeks = sum(1 for score in weekly_scores if score >= boom_threshold)
        bust_weeks = sum(1 for score in weekly_scores if score <= bust_threshold)
        total_weeks = len(weekly_scores)
        
        boom_rate = boom_weeks / total_weeks
        bust_rate = bust_weeks / total_weeks
        
        # Calculate variance (coefficient of variation)
        mean = sum(weekly_scores) / total_weeks
        variance = sum((x - mean) ** 2 for x in weekly_scores) / total_weeks
        std_dev = variance ** 0.5
        
        # Consistency score (inverse of coefficient of variation)
        coefficient_of_variation = std_dev / mean if mean > 0 else 0
        consistency_score = 1 / (1 + coefficient_of_variation)  # Normalize 0-1
        
        profile = BoomBustProfile(
            player_id=player_id,
            boom_rate=boom_rate,
            bust_rate=bust_rate,
            consistency_score=consistency_score,
            variance=variance
        )
        
        logger.debug(f'Boom/bust profile for {player_id}: boom={boom_rate:.2%}, bust={bust_rate:.2%}')
        return profile
        
    except Exception as error:
        logger.error(f'Failed to calculate boom/bust for {player_id}', exc_info=True)
        return BoomBustProfile(
            player_id=player_id,
            boom_rate=0.0,
            bust_rate=0.0,
            consistency_score=0.0,
            variance=0.0
        )
